// liteserver/example/config.js
module.exports = {
  /**
   * server port
   * @type {Number}
   */
  port: 55555,
  /**
   * server root
   * @type {String}  equal apache server root
   */
  root: './',
  /**
   * if debug
   * @type {Boolean}
   */
  debug: true,
  /**
   * request timeout
   * @type {Number} unit ms, default 15 seconds
   */
  timeout: 15000,
  /**
   * when run with cluster/pm, maybe we need listen port manually
   */
  cancelListen: false,
  /**
   * max post fields
   * @type {Number} default as same as formidable
   */
  maxPostFields: 100,
  /**
   * max post fieldSize
   * @type {Number} default as same as formidable
   */
  maxPostFieldSize: 1024,
  /**
   * post tmp dir
   * @type {String} default as same as formidable
   */
  postTmpDir: '/tmp',
  /**
   * redirect access domain, security!
   *   if you want access *, set this option equal "*"
   *   like
   *     redirectAccess : "*",
   */
  redirectAccess : {
    'www.taobao.com' : true
  },
  /**
  /**
   * inner tpl engine config
   * =========================
   */
  /**
   * tpl root
   * @type {String}
   */
  tpl: './tpl',
  /**
   * compress html output
   * @type {Boolean} false
   */
  compress: false
};
